packages/yoroi-extension/app/components/wallet/send/WalletSendFormSteps/AddNFTDialog.js

```diff
@@ -0,0 +1,197 @@
+import { observer } from 'mobx-react';
+import { defineMessages, intlShape } from 'react-intl';
+import Dialog from '../../../widgets/Dialog';
+import DialogCloseButton from '../../../widgets/DialogCloseButton';
+import styles from './AddNFTDialog.scss';
+import type { $npm$ReactIntl$IntlFormat } from 'react-intl';
+import {
+  MultiToken,
+} from '../../../../api/common/lib/MultiToken';
+import { genFormatTokenAmount, getTokenStrictName, getTokenIdentifierIfExists, } from '../../../../stores/stateless/tokenHelpers';
+import SearchIcon from '../../../../assets/images/assets-page/search.inline.svg';
+import NoItemsFoundImg from '../../../../assets/images/dapp-connector/no-websites-connected.inline.svg'
+import { truncateToken } from '../../../../utils/formatters';
+import NoNFT from '../../../../assets/images/nft-no.inline.svg'
+import type { FormattedNFTDisplay } from '../../../../utils/wallet'
+import BigNumber from 'bignumber.js'
+import type {
+  TokenLookupKey,
+} from '../../../../api/common/lib/MultiToken';
+import type { TokenRow } from '../../../../api/ada/lib/storage/database/primitives/tables';
+import type { UriParams } from '../../../../utils/URIHandling';
+
+type Props = {|
+  +onClose: void => void,
+  +spendableBalance: ?MultiToken,
+  +classicTheme: boolean,
+  +getTokenInfo: $ReadOnly<Inexact<TokenLookupKey>> => $ReadOnly<TokenRow>,
+  +updateAmount: (?BigNumber) => void,
+  +uriParams: ?UriParams,
+  +selectedToken: void | $ReadOnly<TokenRow>,
+  +validateAmount: (
+    amountInNaturalUnits: BigNumber,
+    tokenRow: $ReadOnly<TokenRow>,
+  ) => Promise<[boolean, void | string]>,
+  +defaultToken: $ReadOnly<TokenRow>,
+|};
+
+type State = {|
+  nftsList: FormattedNFTDisplay[],
+|}
+
+
+
+export const messages: Object = defineMessages({
+  nameAndTicker: {
+    id: 'wallet.assets.nameAndTicker',
+    defaultMessage: '!!!Name and ticker',
+  },
+  quantity: {
+    id: 'wallet.assets.quantity',
+    defaultMessage: '!!!Quantity',
+  },
+  identifier: {
+    id: 'wallet.assets.fingerprint',
+    defaultMessage: '!!!Fingerprint',
+  },
+  search: {
+    id: 'wallet.assets.search',
+    defaultMessage: '!!!Search',
+  },
+  noAssetFound: {
+    id: 'wallet.assets.noAssetFound',
+    defaultMessage: '!!!No Asset Found',
+  },
+  noTokensYet: {
+    id: 'wallet.send.form.dialog.noTokensYet',
+    defaultMessage: '!!!There are no tokens in your wallet yet'
+  },
+  minAda: {
+    id: 'wallet.send.form.dialog.minAda',
+    defaultMessage: '!!!min-ada'
+  },
+  add: {
+    id: 'wallet.send.form.dialog.add',
+    defaultMessage: '!!!add'
+  },
+  nNft: {
+    id: 'wallet.send.form.dialog.nNft',
+    defaultMessage: '!!!NFT ({number})',
+  },
+});
+
+@observer
+export default class AddNFTDialog extends Component<Props, State> {
+
+  static contextTypes: {|intl: $npm$ReactIntl$IntlFormat|} = {
+    intl: intlShape.isRequired,
+  };
+
+  componentDidMount(): void {
+    this.setState({ nftsList: this.genNftsList() })
```

Comment:
Is it really necessary to have this state, given that  the code calls `this.genNftsList()` to compute the list in multiple places?

Reply:
The same question applies to `AddTokenDialog`.

Reply:
I need to know both the `current` and `full` list of `tokens` and `nfts` as I am displaying some stuff depending on both of them. 


To go around all of these function calls. I updated both `AddTokenDialog ` and `AddNFTDialog` so that 

the state has both `current` and `full` tokens/nts list buy doing this I will call `getTokens` only once when the component mounts!